In [53]:
import tensorflow as tf
from tensorflow.keras.preprocessing.text import one_hot


In [54]:
### sentences
voc_size =10000

sent=[ 'the glass of milk',
     'the glass of juice',
     'the cup of tea',
    'I am a good boy',
     'I am a good developer',
     'understand the meaning of words',
     'your videos are good',]

## 1. One-hot Encoding

In [55]:
#Convert to one-hot represantation
one_hot_repr = [one_hot(input_text=words, n = voc_size) for words in sent]
one_hot_repr

[[644, 5644, 7433, 2794],
 [644, 5644, 7433, 9472],
 [644, 1050, 7433, 5765],
 [1132, 2551, 8360, 2982, 6192],
 [1132, 2551, 8360, 2982, 9428],
 [9476, 644, 4738, 7433, 9941],
 [7301, 8165, 7610, 2982]]

10000 talik vocabulary dictionarydagi tepadagi so'zlarning indexsi. Ya'ni "the", "glass" so'zlari 644, 5644 orinda turadi 10000 talik vektorda va 1 tonda 644 orindagi "the" sozi o'zi 1 va qolgan barcha so'zlar 0 bo'ladi. ya'ni (1,10000) vectorda faqat 644 orinda turgan index 1 qolgani 0 boladi

## 2. Word Embedding

In [56]:
from tensorflow.keras.layers import Embedding
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
import numpy as np

Sentence har xil uzunlikda so'zlar bo'ladi. RNN uchun bir xil input bolishi kerak shu sababli pad_seqences orqali bir xil ravishda oldiga yoki orqasiga 0 berib keltirib olamiz

In [57]:
sent_length = 8  
embedded_data = pad_sequences(sequences=one_hot_repr, maxlen=sent_length, padding='pre') #pre-oldidan, post - keyin 
embedded_data

array([[   0,    0,    0,    0,  644, 5644, 7433, 2794],
       [   0,    0,    0,    0,  644, 5644, 7433, 9472],
       [   0,    0,    0,    0,  644, 1050, 7433, 5765],
       [   0,    0,    0, 1132, 2551, 8360, 2982, 6192],
       [   0,    0,    0, 1132, 2551, 8360, 2982, 9428],
       [   0,    0,    0, 9476,  644, 4738, 7433, 9941],
       [   0,    0,    0,    0, 7301, 8165, 7610, 2982]])

In [58]:
#Feature represantation 
fdim = 10 #har bir soz 10 talik fazoda qiymatga ega boladi. Word2Vec 

In [61]:
model = Sequential()
model.add(Embedding(input_dim=voc_size, output_dim=fdim,input_length=sent_length))
model.compile('adam', 'mse')
model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_3 (Embedding)      (None, 8, 10)             100000    
Total params: 100,000
Trainable params: 100,000
Non-trainable params: 0
_________________________________________________________________


In [62]:
y = model.predict(embedded_data)
y

array([[[-1.23054869e-02, -1.88697930e-02, -2.97861584e-02,
          4.32480015e-02, -5.34498692e-03,  1.14791766e-02,
          2.20578946e-02, -3.72551084e-02, -2.19354164e-02,
          3.00906189e-02],
        [-1.23054869e-02, -1.88697930e-02, -2.97861584e-02,
          4.32480015e-02, -5.34498692e-03,  1.14791766e-02,
          2.20578946e-02, -3.72551084e-02, -2.19354164e-02,
          3.00906189e-02],
        [-1.23054869e-02, -1.88697930e-02, -2.97861584e-02,
          4.32480015e-02, -5.34498692e-03,  1.14791766e-02,
          2.20578946e-02, -3.72551084e-02, -2.19354164e-02,
          3.00906189e-02],
        [-1.23054869e-02, -1.88697930e-02, -2.97861584e-02,
          4.32480015e-02, -5.34498692e-03,  1.14791766e-02,
          2.20578946e-02, -3.72551084e-02, -2.19354164e-02,
          3.00906189e-02],
        [-2.42533218e-02, -3.14685255e-02, -2.76099443e-02,
         -1.31262913e-02, -3.98180857e-02,  2.21918859e-02,
         -2.87373792e-02, -1.06835254e-02, -1.095776

In [43]:
y.shape
# 7,8,10 : 
# 7 - ta gap bor edi shuning uchun 7 ta 
# 8 - bu sent lenth of each sentence. 7 ta ni ichidagi har bir gap ichidagi sozlar soni 
# 10 - dimension feature dimensiondagi har bir sozni 

(7, 8, 10)

In [51]:
print(embedded_data[0])
model.predict(embedded_data[0])

[   0    0    0    0  644 5644 7433 2794]


array([[[-0.02115908,  0.00768881,  0.01480376,  0.0166732 ,
         -0.01184358,  0.00553504, -0.00033022, -0.04106849,
         -0.01271313,  0.03624954]],

       [[-0.02115908,  0.00768881,  0.01480376,  0.0166732 ,
         -0.01184358,  0.00553504, -0.00033022, -0.04106849,
         -0.01271313,  0.03624954]],

       [[-0.02115908,  0.00768881,  0.01480376,  0.0166732 ,
         -0.01184358,  0.00553504, -0.00033022, -0.04106849,
         -0.01271313,  0.03624954]],

       [[-0.02115908,  0.00768881,  0.01480376,  0.0166732 ,
         -0.01184358,  0.00553504, -0.00033022, -0.04106849,
         -0.01271313,  0.03624954]],

       [[ 0.04486108, -0.04242301,  0.02601459, -0.02597016,
         -0.00330074,  0.00747262,  0.01275042, -0.01248965,
         -0.03469765,  0.03023437]],

       [[-0.0029882 , -0.01001918, -0.03472742, -0.00932752,
          0.00476678, -0.03146588, -0.0154718 ,  0.03631573,
         -0.02749444, -0.02310778]],

       [[-0.00590706,  0.01294312,  0.02